### PR TITLE
Fix/div zero

### PR DESCRIPTION
**Pull request by 0xDmtri**
## Motivation

Fix annoying bug related to not enough bits that results in div zero due to very small price number.

## Solution

1. Convert to Q128 before sqrtPrice.
2. Add two forge tests.

## PR Checklist

- [+] Added Tests
- [-] Added Documentation
- [-] Breaking changes

### CODE REVIEW

File: contracts/src/GetWethValueInPoolBatchRequest.sol [hunk 1]
```diff
@@ -256,7 +256,7 @@ contract GetWethValueInPoolBatchRequest {
             ///@notice The first value is a Q96 representation of p_token0, the second is 128X fixed point representation of p_token1.
             uint256 priceSquaredShiftQ96 = token0IsReserve0
                 ? priceSquaredX96 / Q96
-                : (Q96 * 0xffffffffffffffffffffffffffffffff) / (priceSquaredX96 / Q96);
+                : (Q96 * 0xffffffffffffffffffffffffffffffff) / (priceSquaredX96 << 128 / Q96);
```
Comment (0xOsiris) [inline contracts/src/GetWethValueInPoolBatchRequest.sol:259]: You need to change the numerator here. You are trying to invert the price. Your denominator is already Q128 after doing `<< 128`, so you want to do 1/ price. Since it's already in Q128 I believe this should be 

```solidity
uint256 priceSquaredShiftQ96 = token0IsReserve0
                ? priceSquaredX96 / Q96
                : type(uint256).max / (priceSquaredX96 << 128 / Q96);
```

**Review by 0xOsiris — CHANGES_REQUESTED** [review]

File: contracts/src/GetWethValueInPoolBatchRequest.sol [hunk 2]
```diff
@@ -256,7 +256,7 @@ contract GetWethValueInPoolBatchRequest {
             ///@notice The first value is a Q96 representation of p_token0, the second is 128X fixed point representation of p_token1.
             uint256 priceSquaredShiftQ96 = token0IsReserve0
                 ? priceSquaredX96 / Q96
```
Comment (0xOsiris) [inline contracts/src/GetWethValueInPoolBatchRequest.sol:258]: Also is `Q96 > priceSquaredX96` here ever?

File: contracts/src/GetWethValueInPoolBatchRequest.sol (hunk 1 above)
Comment (0xDmtri) [reply contracts/src/GetWethValueInPoolBatchRequest.sol:259]: Ah yes, my bad. Thanks!

File: contracts/src/GetWethValueInPoolBatchRequest.sol (hunk 2 above)
Comment (0xDmtri) [reply contracts/src/GetWethValueInPoolBatchRequest.sol:258]: i havent found the case when this happens to test it tho(( But im sure it can happen. What do you think we should do?

Comment (0xDmtri) [reply contracts/src/GetWethValueInPoolBatchRequest.sol:258]: ```(priceSquaredX96 << 32) >> 64``` ?

Comment (0xOsiris) [reply contracts/src/GetWethValueInPoolBatchRequest.sol:258]: do `priceSquaredX96 << 128 / Q96`. Then below 

change 

```solidity
priceX128 = token0IsReserve0
                ? (uint256(priceSquaredShiftQ96) * 0xffffffffffffffffffffffffffffffff) / Q96
                : priceSquaredShiftQ96;
```

to 
```solidity
priceX128 = token0IsReserve0
                ? priceSquaredShiftQ96 / Q96
                : priceSquaredShiftQ96;
```

Comment (0xOsiris) [reply contracts/src/GetWethValueInPoolBatchRequest.sol:258]: We should really add some unit tests on this function after these changes. Maybe we could separate all peripheral functions into a deployable inherited contract that we can then easily unit test in foundry

**Comment by 0xDmtri** [comment]
"We should really add some unit tests on this function after these changes. Maybe we could separate all peripheral functions into a deployable inherited contract that we can then easily unit test in foundry"

Just for this contract? Should we do it in this pull request or make a new one? U wanna make an abstract contract right?

**Comment by 0xDmtri** [comment]
Another edge case was found:

It will still produce a div 0 error if `sqrtPriceX96` == 0. I dont think we can fix that except of doing:

`require(sqrtPriceX96 != 0)`

@0xOsiris thoughts?

**Comment by 0xDmtri** [comment]
Change log:

1. Added another test case in Foundry.
2. Fixed bug with Uni V3 pools when `sqrtPriceX96 == 0`.
3. Fixed bug with Uni V3 pools when Token does not impl `decimals()`.
4. Added weth value test in Rust.

**Comment by 0xOsiris** [comment]
> "We should really add some unit tests on this function after these changes. Maybe we could separate all peripheral functions into a deployable inherited contract that we can then easily unit test in foundry"
> 
> Just for this contract? Should we do it in this pull request or make a new one? U wanna make an abstract contract right?

Yeah, we can isolate this into a different PR though

**Review by 0xOsiris — APPROVED** [review]
Nice work on this! LGTM

**Comment by 0xDmtri** [comment]
ah shit, theres a conflict with .gitignore, will resolve now

**Comment by 0xDmtri** [comment]
done!
